### DISCUSSION TITLE
Adding Async Partial Methods that can return a awaitable Task.

### DISCUSSION THREAD

**Discussion by MicrosoftHawk** [dotnet/csharplang#7076] [Language Ideas]
I would like to see the posibbilty of using Async Partial Methods that can return a awaitable Task.

I have generated code but sometimes I want to do something slightly different. I then have the possibility of using a partial method however this cant return an awaitable async Task and then I have to use syncrhonous methods or do strange tweaks to work with the rest of my async code.

**Comment by Joe4evr** [comment]
You're over 2 years behind the times. As of C# 9, as part of the effort to support more versatile generated code scenarios, [partial methods are no longer limited to return `void`](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/extending-partial-methods), so there's nothing stopping you from declaring `partial Task SomeOperationAsync();` with an implementation filled in by a generator and that's perfectly awaitable.

**Reply by MicrosoftHawk** [reply]
Sorry, I think my question was wrong.

I want the implementation of the partial method to be optional and that still seems to be only possible with the returntype void.
The generated code is fine in most of the scenarios but sometimes I want to be able to add some custom code.

As a possible option the c# compiler  could just create a method that only returns a completed Task if there isn't an implementation of the partial method.

Maybe I am missing something.

**Reply by kjkrum** [reply]
@MicrosoftHawk I think you could implement this with an incremental source generator. I don't have a fully working solution, but I've been playing around and I think I have a workable concept.

It's helpful to understand how source generators work. First your hand-written source is compiled, then source generators run in parallel against the compilation, then your hand-written source is compiled again with the output of the generators. Any diagnostics from the first compilation pass are available to the generator. So your generator could look for CS8795, use the diagnostic location to find the corresponding syntax node, and determine whether it's suitable for generating a default implementation. If the declaration and generated implementation match, there's no error in the second compilation pass. If the target project already provides an implementation, there will be no CS8795 in the first pass and nothing will be generated.

**Reply by TahirAhmadov** [reply]
Why can't your source generator create the `return Task.CompletedTask;` implementation? Presumably, there is a condition which triggers the creation of your custom implementation, just throw the aforementioned "filler" in the "else" branch of your logic.

**Reply by MicrosoftHawk** [reply]
Thank you for your suggestions. I posted another discussion where there are a lot more nice suggestions.
https://github.com/dotnet/csharplang/discussions/7083

I ended up with a workaround but I still feel a partial method should also be able to return an awaitable task withoud adding the complexity of more generated code. 

It would make te language better.
